### PR TITLE
Lazy containers

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: nx/lib/nx/lazy_container.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  Sometimes building tensors for a container is an expensive
+  operation, so we want to allow that to happen lazily.
+
+  This module provides a single traverse implementation
+  that emits the tensor template and a function that computes
+  the tensor as two distinct values. Then a tensor is only
+  allocated if necessary.
+
+  This protocol is used throughout `Nx.Defn` API. This means
+  compilation, jitting, and streaming will only realized lazy
```
Comment (jonatanklosko) [inline nx/lib/nx/lazy_container.ex:14]:
```suggestion
  compilation, jitting, and streaming will only realize lazy
```

File: nx/lib/nx/lazy_container.ex [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    map
+    |> Map.to_list()
+    |> Enum.sort()
+    |> Enum.map_reduce(acc, fn {k, v}, acc ->
+      {v, acc} = Nx.Defn.Composite.lazy_traverse(v, acc, fun)
+      {{k, v}, acc}
+    end)
+    |> then(fn {list, acc} -> {Map.new(list), acc} end)
+  end
+end
```
Comment (jonatanklosko) [inline nx/lib/nx/lazy_container.ex:64]: Are these just to avoid the `Nx.Container` impl lookup?

File: nx/lib/nx/lazy_container.ex [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  the tensor as two distinct values. Then a tensor is only
+  allocated if necessary.
+
+  This protocol is used throughout `Nx.Defn` API. This means
+  compilation, jitting, and streaming will only realized lazy
+  tensors when necessary.
+
+  If a data structure does not implement this protocol,
+  a default implementation that converts eager to lazy is
+  done by default.
```
Comment (jonatanklosko) [inline nx/lib/nx/lazy_container.ex:19]:
```suggestion
  a default implementation converts eager to lazy using
  `Nx.Container`.
```

File: nx/lib/nx/lazy_container.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx/lazy_container.ex:64]: Correct! I will add a note!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**